### PR TITLE
feat: custom base job name for jumpstart models/estimators

### PR DESCRIPTION
**Pull request by evakravi**
*Issue #, if available:*

*Description of changes:*
Adds custom base job name for jumpstart models/estimators (`sagemaker-jumpstart` prefix). In addition, this PR fixes an integration test not tagging a `Model` so that it can be flagged for deletion. 

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: doc/overview.rst [hunk 1]
```diff
@@ -773,11 +773,10 @@ Deployment may take about 5 minutes.
        instance_type=instance_type,
    )
 
-Because ``catboost`` and ``lightgbm`` rely on the PyTorch Deep Learning Containers
-image, the corresponding Models and Endpoints display the “pytorch”
-prefix when viewed in the AWS console. To verify that these models
-were created successfully with your desired base model, refer to
-the ``Tags`` section.
+Because the model and script URIs are owned by JumpStart, the endpoint,
```
Comment (JGuinegagne) [inline doc/overview.rst:776]: I am not a fan of 'are owned by JumpStart' in public documentation. Besides, isn't it and/or, ie if a customer only uses their own script with a model URIs from JumpStart, would the job use this prefix? How about 'are distributed by SageMaker JumpStart' ?

File: src/sagemaker/estimator.py [hunk 2]
```diff
@@ -570,7 +571,9 @@ def _ensure_base_job_name(self):
         """Set ``self.base_job_name`` if it is not set already."""
         # honor supplied base_job_name or generate it
         if self.base_job_name is None:
-            self.base_job_name = base_name_from_image(self.training_image_uri())
+            self.base_job_name = get_jumpstart_base_name_if_jumpstart_model(
```
Comment (JGuinegagne) [inline src/sagemaker/estimator.py:574]: non-blocking: I think the implementation of line 1215 is more readable, so to be consistent I would prefer:
```py
jumpstart_base_name = get_jumpstart_base_name_if_jumpstart_model(...)
self.base_job_bame = jumpstart_base_name or base_name_from_image(self.training_image_uri())
```

File: src/sagemaker/jumpstart/utils.py [hunk 3]
```diff
@@ -232,6 +232,19 @@ def add_single_jumpstart_tag(
     return curr_tags
 
 
+def get_jumpstart_base_name_if_jumpstart_model(
+    *uris,
```
Comment (JGuinegagne) [inline src/sagemaker/jumpstart/utils.py:236]: could you refine the typing please, ie would one of these work `*uris: List[Optional[str]]` or `*uris: Sequence[Optional[str]]` ?

File: src/sagemaker/model.py [hunk 4]
```diff
@@ -529,10 +531,14 @@ def _create_sagemaker_model(self, instance_type=None, accelerator_type=None, tag
             tags=tags,
         )
 
-    def _ensure_base_name_if_needed(self, image_uri):
+    def _ensure_base_name_if_needed(self, image_uri, script_uri, model_uri):
         """Create a base name from the image URI if there is no model name provided."""
```
Comment (JGuinegagne) [inline src/sagemaker/model.py:535]: Could you please update the docstring to explain the JumpStart-specific logic?

File: src/sagemaker/model.py [hunk 5]
```diff
@@ -529,10 +531,14 @@ def _create_sagemaker_model(self, instance_type=None, accelerator_type=None, tag
...
         if self.name is None:
-            self._base_name = self._base_name or utils.base_name_from_image(image_uri)
+            self._base_name = (
+                self._base_name
+                or get_jumpstart_base_name_if_jumpstart_model(script_uri, model_uri)
+                or utils.base_name_from_image(image_uri)
+            )
```
Comment (JGuinegagne) [inline src/sagemaker/model.py:541]: non-blocking: I would prefer to define a `jumpstart_base_name` variable here as well.

File: tests/unit/sagemaker/jumpstart/test_utils.py [hunk 6]
```diff
@@ -874,3 +875,11 @@ def make_deprecated_spec(*largs, **kwargs):
             "pytorch-eqa-bert-base-cased",
             "*",
         )
+
+
+def test_get_jumpstart_base_name_if_jumpstart_model():
+    uris = [random_jumpstart_s3_uri("random_key") for _ in range(random.randint(1, 10))]
+    assert JUMPSTART_RESOURCE_BASE_NAME == utils.get_jumpstart_base_name_if_jumpstart_model(*uris)
```
Comment (JGuinegagne) [inline tests/unit/sagemaker/jumpstart/test_utils.py:882]: nit: could you add tests where (a) only the last item of the args is a jumpstart s3 URI, (b) an item in the middle of the array is a jumpstart s3 URI?

**Review by JGuinegagne — CHANGES_REQUESTED** [review]
Just a few clean-up and optional comments please.

**Review by JGuinegagne — APPROVED** [review]

File: src/sagemaker/jumpstart/utils.py [hunk 7]
```diff
@@ -232,6 +232,19 @@ def add_single_jumpstart_tag(
     return curr_tags
 
 
+def get_jumpstart_base_name_if_jumpstart_model(
+    *uris: Optional[str],
+) -> Optional[str]:
+    """Return default JumpStart base name if a URI belongs to JumpStart.
+
+    If no URIs belong to JumpStart, return None.
```
Comment (mufaddal-rohawala) [inline src/sagemaker/jumpstart/utils.py:240]: Can we add Args in docstrings.

**Review by jeniyat — APPROVED** [review]

File: src/sagemaker/jumpstart/utils.py (hunk 7 above)
Comment (navinsoni) [inline src/sagemaker/jumpstart/utils.py:240]: Can you please address Mufis' comment

**Review by navinsoni — CHANGES_REQUESTED** [review]

File: src/sagemaker/jumpstart/utils.py (hunk 7 above)
Comment (evakravi) [reply src/sagemaker/jumpstart/utils.py:240]: I did (`*uris (Optional[str]): URI to test for association with JumpStart.`). I believe you're looking at an old version.

**Review by navinsoni — APPROVED** [review]

**Pull request merged**